### PR TITLE
761 fix dense_slices issue for sliding window

### LINKED ISSUE #761: Incorrect operation of dense_patch_slices
**Describe the bug**
I found that the sliding window inference function was much slower than expected. Looking into this, I found that the dense_patch_slices function is returning many duplicated slices for certain inputs.

**To Reproduce**
For example:

```
>>> x = dense_patch_slices(image_size=(100,100), patch_size=(50,50), scan_interval=(25,25))
>>> for slice in x:
...     print(slice)
... 
(slice(0, 50, None), slice(0, 50, None))
(slice(0, 50, None), slice(25, 75, None))
(slice(0, 50, None), slice(50, 100, None))
(slice(0, 50, None), slice(50, 100, None))
(slice(25, 75, None), slice(0, 50, None))
(slice(25, 75, None), slice(25, 75, None))
(slice(25, 75, None), slice(50, 100, None))
(slice(25, 75, None), slice(50, 100, None))
(slice(50, 100, None), slice(0, 50, None))
(slice(50, 100, None), slice(25, 75, None))
(slice(50, 100, None), slice(50, 100, None))
(slice(50, 100, None), slice(50, 100, None))
(slice(50, 100, None), slice(0, 50, None))
(slice(50, 100, None), slice(25, 75, None))
(slice(50, 100, None), slice(50, 100, None))
(slice(50, 100, None), slice(50, 100, None))
```

**Expected behavior**
dense_patch_slices should return a list of slice objects which cover the volume, overlap by the specified scan interval, but without any duplicates.

**Environment (please complete the following information):**
 - OS: Centos 7
 - MONAI version: 0.2.0+34.g267407f

**Additional context**
I would like to fix this problem, and add a few unit tests for this function at the same time.

### PR DESCRIPTION
**Pull request by Nic-Ma**
Fixes #761 .

### Description
This PR fixed the duplicated slices issue in sliding window

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/utils.py [hunk 1]
```diff
@@ -103,10 +103,17 @@ def dense_patch_slices(
...
+    scan_num = list()
+    for i in range(num_spatial_dims):
+        if scan_interval == 0:
+            scan_num.append(1)
+        else:
+            num = int(math.ceil(float(image_size[i]) / scan_interval[i]))
+            for d in range(num):
+                if d * scan_interval[i] + patch_size[i] >= image_size[i]:
+                    break
+            scan_num.append(d + 1)
```
Comment (ericspod) [inline monai/data/utils.py:115]: I would use a loop variable outside of the loop, Python allows it but it's not standard practice and a little hard to read. The whole for loop can be made into an expression however:

```
scan_dim = first(d for d in range(num) if d * scan_interval[i] + patch_size[i] >= image_size[i])
scan_num.append(scan_dim + 1)
```

File: monai/data/utils.py [hunk 2]
```diff
@@ -103,10 +103,17 @@ def dense_patch_slices(
     patch_size = get_valid_patch_size(image_size, patch_size)
     scan_interval = ensure_tuple_size(scan_interval, num_spatial_dims)
 
-    scan_num = [
-        int(math.ceil(float(image_size[i]) / scan_interval[i])) if scan_interval[i] != 0 else 1
-        for i in range(num_spatial_dims)
-    ]
+    scan_num = list()
+    for i in range(num_spatial_dims):
+        if scan_interval == 0:
```
Comment (ericspod) [inline monai/data/utils.py:108]: Should this not be `if scan_interval[i] == 0:` ?

File: monai/data/utils.py (hunk 1 above)
Comment (Nic-Ma) [reply monai/data/utils.py:115]: updated in the latest commit.
Thanks.

File: monai/data/utils.py (hunk 2 above)
Comment (Nic-Ma) [reply monai/data/utils.py:108]: Yes, sorry for my typo.
Updated in the latest commit.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Thanks for your review!
I updated the code according to your comments.
Could you please help review it again?

Thanks.

**Comment by ericspod** [comment]
I think it's fine, when good to go we can merge it in.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

I tried to enable integration tests for this branch.
As you can see, it failed at:
```
FAIL: test_training (tests.test_integration_sliding_window.TestIntegrationSlidingWindow)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/__w/MONAI/MONAI/tests/test_integration_sliding_window.py", line 85, in test_training
    np.testing.assert_allclose(np.sum(output_image), 33583)
```
Actually, I don't know why but it's not the latest code...I already updated to `33621` in the latest commit.
And I ran the integration tests locally, it passed.

So can we remove this branch from test and merge it directly?

Thanks.

**Comment by Nic-Ma** [comment]
The `test_integration_sliding_window.TestIntegrationSlidingWindow` test is not only for master branch.
So it passed in CI now.
Thaks.

**Pull request merged**